### ISSUE TITLE
Incorrect bounce angle of RididBody2d

### ISSUE THREAD

**Issue by GaidamakUA** [godotengine/godot#32138]
**Godot version:**
v3.1.1.stable.official
**OS/device including version:**
Mac OS X and Linux Ubuntu
**Issue description:**
RigidBody2d collides with flat StaticBody2d and bounces with random angle. Often, with the same it came from. Happens on relatively high speed of RigidBody2d
**Steps to reproduce:**
1. Run the project
2. Bump the RigidBody2d with KinematicBody2d so it's speed increases
3. Observe the bug.

Here are the description with video and screenshots attached:
https://www.reddit.com/r/godot/comments/d4455z/bug_i_godot_physics_system/
**Minimal reproduction project:**
https://github.com/GaidamakUA/GodotExperiments/tree/master/arcanoid

**Comment by homer666** [comment]
Setting "Continuous Cd" on the Ball to "Cast Shape" seems to make it behave correctly. I couldn't tell you why though.

**Comment by GaidamakUA** [comment]
Yes, I can confirm, it works with "Cast shape". Thanks. Just to give you a bit more context. I've tried to implement borders in previous version by collision shape in TileMap. And it behaved in a similar manner. And "Cast Shape" didn't work. You need me to make another sample project for that?

**Comment by homer666** [comment]
@GaidamakUA Yeah that would definitely help, if it isn't too much trouble.

**Comment by GaidamakUA** [comment]
It's pretty same but collision is implemented via tilest
https://github.com/GaidamakUA/GodotExperiments/tree/master/arcanoid_tileset_collision
For this version Cast Shape doesn't fix the issue.

**Comment by Feniks-Gaming** [comment]
https://www.reddit.com/r/godot/comments/eerh7a/why_is_my_bouncing_inconsistent/

I have the same issues in my project using 3.1

**Comment by maximinus** [comment]
I have exactly the same issue. I would guess the object is travelling inside the static body and this is causing it to lose accuracy somewhere. I have seen this on Windows 10 and Linux. The issue can be reduced somewhat with a slower moving object and "Cast shape" a little better but cannot remove it totally.

**Comment by Feniks-Gaming** [comment]
I can confirm what @maximinus is saying using a square rather than circle collision shape, using cast shape collision reduces this significantly but doesn't fix the issue fully. As result if one was to create a bouncing ball that goes down and bounces up on collision it would eventually start bounding other angle and at different speed despite no other forces being used. 

What is worse [documentation](https://docs.godotengine.org/en/3.1/tutorials/physics/physics_introduction.html#using-rigidbody2d) suggest that RidigBody2d should be exatly what we are supposed to use for simple games where we want ball to collide with some blocks.

**Comment by jtaart** [comment]
The problem seemed to be located here (collision detection by raycasting mode):

https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/servers/physics_2d/body_pair_2d_sw.cpp#L171


The contact seems to be flipped. Don't exactly know whats going on with the function, but it seems a little off.  
https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/servers/physics_2d/body_pair_2d_sw.cpp#L214-L217

[flipped gif](https://imgur.com/a/UsgoR5z)
[unflipped gif](https://imgur.com/a/fuN2RIX)
[temp fix gif](https://imgur.com/a/hEq0tXg)

**Comment by jtaart** [comment]
The contact seems to be flipped. Don't exactly know whats going on with the function, but it seems a little off.  
https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/servers/physics_2d/body_pair_2d_sw.cpp#L214-L217

[flipped gif](https://imgur.com/a/UsgoR5z)
[unflipped gif](https://imgur.com/a/fuN2RIX)
[temp fix gif](https://imgur.com/a/hEq0tXg)

**Comment by madmiraal** [comment]
This may be related to #28438. It appears to occur when colliding with the point between two `CollisionShape`s. Since it calculates the collisions separately, there is a collision with the corner of one shape i.e. a little bit of the edge between the shapes.

**Comment by Yolwoocle** [comment]
This still seems to happen in 4.2.1. Is there any plan for a fix? It's been 4 years.
